CLI Job validate SIGSEGV with job type=system and migrate stanza

### Nomad version
Nomad v0.9.0-rc1 (7c00ab4f3f37cfd1e258b38fd2ad99e7bc23e4c3)

Affects 0.8.7 at least as well

### Operating system and Environment details
Darwin 18.2, Linux, etc

### Issue
A job file with `type = "system"` and containing a `job:migrate{}` stanza causes `nomad validate <job>.nomad` to panic if a group is defined.

Change the type to `service` and `validate` successfully runs, complaining about missing tasks.

### Reproduction steps
Create job files from below.


Validate job file with `service` scheduler:
```
~/d/nomad> ./nomad job validate panic.nomad 
Job validation errors:
2 error(s) occurred:

* Missing job datacenters
* Task group panicgrp validation failed: 1 error(s) occurred:

* Missing tasks for task group
```

Validate job file with `system` scheduler:
```
~/d/nomad> ./nomad job validate panic.nomad 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x53276f9]

goroutine 1 [running]:
github.com/hashicorp/nomad/command/agent.ApiTgToStructsTG(0xc0002e5540, 0xc0000ca4d0)
        /tmp/nomad-workspace/2019-03-21-1553195760/gopath/src/github.com/hashicorp/nomad/command/agent/job_endpoint.go:691 +0x5c9
github.com/hashicorp/nomad/command/agent.ApiJobToStructJob(0xc0003fa000, 0xc00043a180)
        /tmp/nomad-workspace/2019-03-21-1553195760/gopath/src/github.com/hashicorp/nomad/command/agent/job_endpoint.go:656 +0x45f
github.com/hashicorp/nomad/command.(*JobValidateCommand).validateLocal(0xc000128370, 0xc0003fa000, 0x0, 0xc0001e4f80, 0x0)
        /tmp/nomad-workspace/2019-03-21-1553195760/gopath/src/github.com/hashicorp/nomad/command/job_validate.go:120 +0x5d
github.com/hashicorp/nomad/command.(*JobValidateCommand).Run(0xc000128370, 0xc00003a1b0, 0x1, 0x1, 0xc00030e1b0)
        /tmp/nomad-workspace/2019-03-21-1553195760/gopath/src/github.com/hashicorp/nomad/command/job_validate.go:85 +0x69a
github.com/hashicorp/nomad/vendor/github.com/mitchellh/cli.(*CLI).Run(0xc00027a640, 0xc00027a640, 0xc0001d4560, 0x3b)
        /tmp/nomad-workspace/2019-03-21-1553195760/gopath/src/github.com/hashicorp/nomad/vendor/github.com/mitchellh/cli/cli.go:255 +0x207
main.RunCustom(0xc00003a190, 0x3, 0x3, 0x0)
        /tmp/nomad-workspace/2019-03-21-1553195760/gopath/src/github.com/hashicorp/nomad/main.go:131 +0x48e
main.Run(0xc00003a190, 0x3, 0x3, 0xc00009e058)
        /tmp/nomad-workspace/2019-03-21-1553195760/gopath/src/github.com/hashicorp/nomad/main.go:76 +0x3f
main.main()
        /tmp/nomad-workspace/2019-03-21-1553195760/gopath/src/github.com/hashicorp/nomad/main.go:72 +0x62
```

### Job file (if appropriate)
*panic.nomad*

```
job "panic" {
    type = "system"
    migrate {}
    group "panicgrp" {}
}
```

*nopanic.nomad*

```
job "panic" {
    type = "service"
    migrate {}
    group "panicgrp" {}
}
```

--- a/api/tasks.go
+++ b/api/tasks.go
@@ -563,13 +563,11 @@ func (g *TaskGroup) Canonicalize(job *Job) {
 	}
 
 	// Merge with default reschedule policy
-	if *job.Type == "service" {
-		defaultMigrateStrategy := &MigrateStrategy{}
-		defaultMigrateStrategy.Canonicalize()
-		if g.Migrate != nil {
-			defaultMigrateStrategy.Merge(g.Migrate)
-		}
-		g.Migrate = defaultMigrateStrategy
+	if g.Migrate == nil && *job.Type == "service" {
+		g.Migrate = &MigrateStrategy{}
+	}
+	if g.Migrate != nil {
+		g.Migrate.Canonicalize()
 	}
 
 	var defaultRestartPolicy *RestartPolicy

--- a/command/agent/job_endpoint_test.go
+++ b/command/agent/job_endpoint_test.go
@@ -15,6 +15,7 @@ import (
 	"github.com/hashicorp/nomad/nomad/structs"
 	"github.com/kr/pretty"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
 func TestHTTP_JobsList(t *testing.T) {
@@ -2033,4 +2034,47 @@ func TestJobs_ApiJobToStructsJob(t *testing.T) {
 	if diff := pretty.Diff(expectedSystemJob, systemStructsJob); len(diff) > 0 {
 		t.Fatalf("bad:\n%s", strings.Join(diff, "\n"))
 	}
+}
+
+// TestHTTP_JobValidate_SystemMigrate asserts that a system job with a migrate
+// stanza fails to validate but does not panic (see #5477).
+func TestHTTP_JobValidate_SystemMigrate(t *testing.T) {
+	t.Parallel()
+	httpTest(t, nil, func(s *TestAgent) {
+		// Create the job
+		job := &api.Job{
+			Region:      helper.StringToPtr("global"),
+			Datacenters: []string{"dc1"},
+			ID:          helper.StringToPtr("systemmigrate"),
+			Name:        helper.StringToPtr("systemmigrate"),
+			TaskGroups: []*api.TaskGroup{
+				{Name: helper.StringToPtr("web")},
+			},
+
+			// System job...
+			Type: helper.StringToPtr("system"),
+
+			// ...with an empty migrate stanza
+			Migrate: &api.MigrateStrategy{},
+		}
+
+		args := api.JobValidateRequest{
+			Job:          job,
+			WriteRequest: api.WriteRequest{Region: "global"},
+		}
+		buf := encodeReq(args)
+
+		// Make the HTTP request
+		req, err := http.NewRequest("PUT", "/v1/validate/job", buf)
+		require.NoError(t, err)
+		respW := httptest.NewRecorder()
+
+		// Make the request
+		obj, err := s.Server.ValidateJobRequest(respW, req)
+		require.NoError(t, err)
+
+		// Check the response
+		resp := obj.(structs.JobValidateResponse)
+		require.Contains(t, resp.Error, `Job type "system" does not allow migrate block`)
+	})
 }

